### PR TITLE
Add ReLU and LeakyReLU activation function in ml module

### PR DESCRIPTION
**Pull request by LaurentBerger**
**Merge with extra**: https://github.com/opencv/opencv_extra/pull/409

Add ReLU and LeakyReLU activation function in ml module

### CODE REVIEW

**Comment by alalek** [comment]
@LaurentBerger Please add tests for new functionality.

**Comment by LaurentBerger** [comment]
@alalek I try to understand what is a test.
In opencv_extra/testdata/ml/legacy there is a [Neural Network](https://github.com/opencv/opencv_extra/blob/master/testdata/ml/legacy/ann_waveform.xml) trains with [data](https://github.com/opencv/opencv_extra/blob/master/testdata/ml/waveform.data).
I think I must train a NN using same data with RELU and LeakyRelu and writ NN in opencv_extra. is it right?

**Comment by alalek** [comment]
I'm assume something like these steps ("/" below means separation between two modes: testdata generation mode / algorithm verification (normal) mode): 
- create simple model (or reuse existed ANN model if it is applicable) with new activation functions
- generate train data + save to opencv_extra (or nothing in case of reusing existed) / load train data from opencv_extra
- model.train()
- save train result to opencv_extra / load reference train result from opencv_extra and verify (somehow) trained result from previous step (norm L2 relative?)
- get reference trained model
- model.predict(inputs  from <train data>)
- save predict() result to opencv_extra / load reference predict() result from opencv_extra and verify (somehow) predicted result from previous step (norm L2 relative?)

Split testdata generation mode and normal verification mode via `#if`:
```
// Uncomment line below to re-generate testdata only.
// Keep line commented to run verification mode.
//#define GENERATE_TESTDATA 
...
result <== model.train(<traindata>)
#ifdef GENERATE_TESTDATA
... save trained "result" model to opencv_extra (model.save(fileStorage) ?) ...
#else
... "reference" = load reference model from opencv_extra ...
... make some checks between "result" model and "reference" ...
#endif
...
```

**Comment by vpisarev** [comment]
:+1:

File: modules/ml/test/test_mltests2.cpp [hunk 1]
```diff
@@ -177,6 +193,62 @@ float ann_calc_error( Ptr<StatModel> ann, Ptr<TrainData> _data, map<int, int>& c
...
+        Ptr<ml::ANN_MLP> y = Algorithm::load<ANN_MLP>(dataname + activationName[i] + ".yml");
+        ASSERT_TRUE(y != NULL) << "Could not load   " << dataname + activationName[i] + ".yml";
+        Mat testSamples = tdata->getTestSamples();
+        Mat rx, ry, dst;
+        x->predict(testSamples, rx);
+        y->predict(testSamples, ry);
+        absdiff(rx, ry, dst);
+        double minVal, maxVal;
+        minMaxLoc(dst, &minVal, &maxVal);
+        ASSERT_TRUE(maxVal<FLT_EPSILON) << "Predict are not equal for " << dataname + activationName[i] + ".yml and " << activationName[i];
```
Comment (alalek) [inline modules/ml/test/test_mltests2.cpp]: BTW, absdiff()+minMaxLoc() => `[cvtest::]norm(rx, ry, NORM_INF)` ([documentation](https://docs.opencv.org/3.3.0/d2/de8/group__core__array.html#gad12cefbcb5291cf958a85b4b67b6149f))

`cvtest` namespace is recommended for test checks.

**Comment by Riyuzakii** [comment]
@LaurentBerger @alalek , Is there some problem with using the relu function currently?
```
    mlp->setActivationFunction(ANN_MLP::SIGMOID_SYM, 0.5, 1);
    mlp->setActivationFunction(ANN_MLP::RELU, 0.5, 1);
```
the first line for sigmoid function works but for relu it says:
```
‘RELU’ is not a member of ‘cv::ml::ANN_MLP’
```
Can you think of any reason as to why this is happening?

**Comment by LaurentBerger** [comment]
ANN_MLP::RELU was introduced in opencv 3.4.0 (Dec 2017)
